AG-12369 Fix keynav using initMenuKeyNav and correct aria roles

olegat opened the pull request:
https://ag-grid.atlassian.net/browse/AG-12369

olegat committed:
AG-12369 Fix keynav using initMenuKeyNav and correct aria roles

alantreadway commented on packages/ag-charts-enterprise/src/features/popover/popover.ts:
```diff
@@ -101,63 +98,15 @@ export class Popover extends _ModuleSupport.BaseModuleInstance implements _Modul
-                    break;
-                case 'Escape':
-                case 'Enter':
-                    opts.onClose();
```
Is `Escape` handling still maintained? I see there is an `onEscape` option to `initMenuKeyNav()` that might need to be used?

olegat replied:
We could easily add an `onEscape: () => this.hide` option, but this wouldn't be a complete solution. The `contextMenu.ts` module keeps a reference to HTMLElement that triggered the popup:

https://github.com/ag-grid/ag-charts/blob/c811b8f9c052098c4dec9b94d76ed508a7583a3f/packages/ag-charts-enterprise/src/features/context-menu/contextMenu.ts#L240

`popover.ts` will need to do something similar. I was going to move the `ContextMenu.lastFocus` feature to be part of the `initMenuKeyNav` method, but on second thoughts I could do this in the PR too.

olegat replied:
Fixed. The `ContextMenu.lastFocus` logic is now part of the `initMenuKeyNav` function and therefore is used by both context menu and financial charts popovers.

alantreadway commented on packages/ag-charts-enterprise/src/features/popover/popover.ts:
```diff
@@ -101,63 +98,15 @@ export class Popover extends _ModuleSupport.BaseModuleInstance implements _Modul
-        };
-        window.addEventListener('mousemove', windowMouseMoveEvent);
-
-        const windowMouseMoveDown = (e: MouseEvent) => {
-            if (e.button !== 0 || popover.contains(e.target as any)) return;
-
-            opts.onClose();
-        };
```
Is the behaviour of 'clicks outside the popup close the popup' still maintained?

olegat replied:
It's working with the Trend Line and Text Annotations menus, but not with the Chart Type menu. I suspect that the Chart Type menu has some other logic that's clashing with.

olegat replied:
Fixed: Clicks outside the popup will close the popup.

Notes:
- Right-clicks will also close the popup (previously, only left-clicks closed the pop-up).
- This change has also been to context menus.

olegat committed:
AG-12369 Rename toolbarManager.ts Event type

olegat committed:
AG-12369 Provide the sourceEvent: MouseEvent to popover.show

This is to allow the module to obtain the `lastFocus` element and restore
its focus when the Escape key is pressed (like with the context menus).

olegat committed:
AG-12369 Make initMenuKeyNav restore lastFocus when ESC is pressed

olegat committed:
AG-12369 Detect mouse clicks downside menus and close them

Putting in logic in keynavUtil.ts is a bit counter-intuitive, because it
was meant to handle keyboard interactions and not mouse interactions. But
as the same, putting this logic in initMenuKeyNav is the best way to
ensure consistency with context menus and financial charts popovers.

We may want to rename this file / functions at some point to make it
clearer that both mouse and keyboard event handlers are initialised.

olegat committed:
AG-12369 Use one implementation of addRemovableEventListener

alantreadway reviewed (approved):
LGTM
